# Import Bibliotek

In [14]:
import pandas as pd
from sklearn import metrics
from sklearn.model_selection import train_test_split
# https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler

# Wczytanie Danych

In [15]:
df_final = pd.read_csv('../data/processed/bank_data_processed.csv', sep=',').copy()

# Model MLP

In [16]:
X = df_final.drop('y', axis=1)
Y = df_final['y']
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

In [17]:
num_cols = ['age', 'balance', 'day', 'campaign', 'pdays', 'previous']
scaler = StandardScaler()

In [18]:
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()
X_train_scaled[num_cols] = scaler.fit_transform(X_train[num_cols])
X_test_scaled[num_cols] = scaler.transform(X_test[num_cols])

In [19]:
mlp = MLPClassifier(hidden_layer_sizes=(20, 10), random_state=0, max_iter=10000)
mlp.fit(X_train_scaled, Y_train)
y_pred = mlp.predict(X_test_scaled)

In [20]:
metrics.confusion_matrix(Y_test, y_pred)

array([[7741,  239],
       [ 770,  293]])

In [21]:
print(metrics.classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.97      0.94      7980
           1       0.55      0.28      0.37      1063

    accuracy                           0.89      9043
   macro avg       0.73      0.62      0.65      9043
weighted avg       0.87      0.89      0.87      9043



# Wnioski

* Brak mechanizmu balansowania wag klas skutkuje to wysokim wskaźnikiem Accuracy, ale niskim Recall (tylko 28% skuteczności w wykrywaniu realnych subskrypcji).
* W przeciwieństwie do Regresji, sieć neuronowa jest bardziej "ostrożna" (wyższy Precision). Jeśli model przewiduje sukces, ma rację w ponad połowie przypadków.
* Model MLP działa jak "czarna skrzynka" = trudniejsze do bezpośredniego wyjaśnienia biznesowego.